rust/src/countme/cookie.rs

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use std::fs;
+use std::io::{BufRead, BufReader};
+use std::path;
+use chrono::prelude::*;
+use std::io::{Result, Write};
+
+/// State directory used to store the countme cookie
+const STATE_DIR: &str = "/var/lib/rpm-ostree";
+/// Cookie file name
+const COUNTME_COOKIE: &str = "countme";
+/// Cookie file format version
+/// Although similar to the DNF Count Me Cookie file version, this is distinct
+/// as this does not include the COUNTME_BUDGET as we will directly query all
+/// repo metalinks once. The file format is:
+/// 
+/// ```
+/// <version> <epoch> <window>
+/// ```
+/// 
+/// Where:
+///   - version: the file format version
+///   - epoch: the timestamp of the first ever counted window
+///   - window: the timestamp of the last successfully counted window
+const COUNTME_VERSION: i64 = 0;
+/// Width of the sliding time window (in seconds): 1 week
+const COUNTME_WINDOW: i64 = 7*24*60*60;
+/// Starting point of the sliding time window relative to the UNIX epoch:
+/// Monday (1970-01-05 00:00:00 UTC)
+/// Allows for aligning the window with a specific weekday
+const COUNTME_OFFSET: i64 = 345600;
+
+pub struct Cookie {
+    epoch: i64,
+    window: i64,
+    now: i64,
+}
+fn parse_cookie_v0(line: &str, now: i64) -> Option<Cookie> {
```

Comment:
If we're freewheeling the cookie format anyway, why not just make it JSON and use serde?

Reply:
That's a good idea. I don't have any strong preference here. I did this mostly to stay close to libdnf existing code. Will try that next.